installers/Types.hs

```diff
@@ -0,0 +1,66 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+module Types
+where
+
+import           Data.Text                           (Text, pack, toLower)
+import           Data.String                         (IsString)
+import           Prelude
+
+
+
+data API
+  = Cardano
+  | ETC
+  deriving (Bounded, Enum, Eq, Read, Show)
+
+data OS
+  = Linux
+  | Macos64
+  | Win64
+  deriving (Eq, Show)
+
+data Cluster
+  = Mainnet
+  | Staging
+  deriving (Bounded, Enum, Eq, Read, Show)
+
+data Config
+  = Launcher
+  | Topology
+  deriving (Eq, Show)
+
+data CI
+  = Appveyor
+  | Travis
+  | Buildkite
+  | Manual
+  deriving (Bounded, Enum, Eq, Read, Show)
+
+data Request
+  = Request
+  { rOS      :: OS
+  , rCluster :: Cluster
+  , rConfig  :: Config
+  } deriving (Eq, Show)
+
+class (Bounded a, Eq a) => Flag a where
```

Comment:
Let's get rid of this magic and rather provide a simple function if we need explicit `CI -> Bool`.

Reply:
Respectfully, I can explain why this is needed and how this works,
  so you can see it is not magic.

  First, the why.

  The Flag type encodes type-safe, semantically enriched boolean values,
  that are not prone to the (somewhat) famous issue of Boolean Blindness.

  If you do not trust my judgement that this is a serious problem, please consider
  referring to these, different sources:

  - https://existentialtype.wordpress.com/2011/03/15/boolean-blindness/
  - https://github.com/quchen/articles/blob/master/algebraic-blindness.md

  Also, you once said yourself, that you're afraid of 'iohk-ops' having all these
  booleans, and the possibility of eventually confusing them in the code, leading
  to things like WipeNodeDBs.

  Indeed, that concern of yours has spurred me to implement the solution you see.

  Next, how it works, and why it is not magic.

  The Bounded type class provides minimum and maximum values for an ADT, whose
  constructors are all 0-ary.  This reference to minimum and maximum allows us to
  encode Bool values naturally and consistently, which gives rise to a type class.

  As you can see, the implementation is rather simple -- and I would argue that it
  serves to solve the stated problem (Boolean Blindness) fairly well.

Reply:
cc @rvl, @jmitchell

Reply:
I share your concerns about boolean blindness, @deepfire. The use of `Bounded` is less obvious to me, though. From [the docs](https://hackage.haskell.org/package/base-4.9.0.0/docs/Prelude.html#t:Bounded):

> The Bounded class may be derived for any enumeration type; minBound is the first constructor listed in the data declaration and maxBound is the last.

The default definitions of `enabled` and `disabled` are the opposite of what I'd expect given that `minBound :: Bool` is `False`. By leaking `Bool` into the ADT the concerns about boolean blindness seep into our logic.

Somewhere we have to implement `Bool -> TestInstaller` because the value depends on the presence or absence of a flag option. As long as we isolate the boolean blindness to this one spot, it's a forgivable and obvious crime. There may be a more elegant way to handle that, but I would prefer we at least get rid of `Bounded`.

Reply:
On top of Jake's arguments, most importantly: we strive to optimize code for reading, not writing. Code is written once and read many, many times in teams.

It took me a couple of minutes to understand how all of this works while having `TestInstaller -> Bool` used inside our functions would still solve boolean blindness and be understandable in a couple of seconds. 

I looked through the code and most of the instances are not used anyway, except in `TestInstaller`. I do not quite understand they would make sense anyway, since those ADTs have more than two items. It's also not obvious that adding a few term level at the beginning or end will change the behavior of the code. If one would implement a simple pattern match, you'd get a compile error - hence I consider such code magic.

Even `TestInstaller` doesn't need `-> Bool` because its only used once and since you don't pass it to a third party library you can do:

```
MacInstaller.hs:  when (oTestInstaller == TestInstaller) $ do
```

eliminating any need for Bounded instance.

Going an extra mile here would be to do proper naming of ADTs, so it's clear from the name what boolean value it implies: `IsTestInstaller = IsTestInstaller | IsNotTestInstaller`.

Reply:
@domenkozar, @jmitchell, interestingly, I find that your suggestions collide somewhat, and so I'm at a certain impasse:
1. Domen suggests a `TestInstaller -> Bool` function, whereas
2. Jake suggests getting rid of `Bool` altogether
We can't do both, so we need some kind of consensus here : -)

The `Bool -> TestInstaller` is supposed to happen as early as possible -- hence the `flag` option parser, which is generic and allows us to avoid mentioning `Bool` in client code entirely.

Note, though, how `flag` depends on ability to _generically negate_ the value -- which depends on the `Flag` type class which defines this negation.  Please note that we can still hide the unsafe `Bool`-involving operations using module boundaries and exports.

@jmitchell, if I understand you correctly, you don't like the `Bounded` dependency of `Flag` (`Ord` is _not_ a dependency of `Flag`, though!) -- but do you have suggestions on how we could otherwise implement negation generically, without involving `Bounded`?
Note, that the `Bool -> Flag` injection needs not be exposed to client code -- we can hide it in an implementation module, if we want the extra safety.

@domenkozar, I'm completely on-board with you with regards to optimization for reading!

But also, do you think that optimization for reading is an argument against abstraction?

Especially abstraction that allows us to deal with the domain language, instead of `Bool` -- with increased safety?

Reply:
We are consistent here with Jake.

Jake is saying we don't need `TestInstaller -> Bool` at all for the moment. I say the same here:

```
Even TestInstaller doesn't need -> Bool because its only used once and since you don't pass it to a third party library 
```

What I was saying about `TestInstaller -> Bool` is that if you DO need it, it won't create boolean blindness since you'll be converting to Bool just before you pass it to `base` package function that requires explicit `Bool`.

@deepfire yes, reading is harder for abstractions, especially leaky ones like encoding Booleans into order of ADTs.

@deepfire please read my previous comment, your implemention is less type safe due to being more implicit.

Reply:
@domenkozar, oh, I don't mind hiding `TestInstaller -> Bool` at all -- they don't need to be anything above an implementation detail.

Done!

Reply:
@deepfire you're not addressing my feedback, please re-read my comments.

Reply:
@domenkozar, I'm sorry that you feel that way -- I'm certainly trying to address your feedback.

Let's go through your previous message bit by bit:

> We are consistent here with Jake.
>
> Jake is saying we don't need TestInstaller -> Bool at all for the moment. I say the same here:
> ```
> Even TestInstaller doesn't need -> Bool because its only used once and since you don't pass it to a third party library 
> ```

The unsafe `TestInstaller -> Bool` is no longer dangerously exposed -- I don't think it's fair to say your feedback wasn't addressed on this point.

> What I was saying about `TestInstaller -> Bool` is that if you DO need it, it won't create boolean blindness since you'll be converting to Bool just before you pass it to base package function that requires explicit Bool. This means your function signatures will still have isTestInstaller and then reading the term-level code you can see it becomes a boolean with an explicity cast.

The new code doesn't need or even suggest using `TestInstaller -> Bool` in the client side, so this section doesn't apply, right?  Any time you need a `Bool`, you can just do `oTestInstaller == TestInstaller`, as was intended all the way.  I think we can agree on that.

So the second part of your previous comment was implicitly addressed.

> yes, reading is harder for abstractions, especially leaky ones like encoding Booleans into order of ADTs.
That simply means assigning interpretation, and it's done once -- where as usage happens multiple times -- so it's a trade off for letting `Bool` not to be used in the user code at all.

And the reason it's needed was already described in detail, previously -- which I'll repeat again, because you didn't seem to recognize this in our discussion:

The need for the `Bounded` instance above -- to enable the generic `flag` mechanism, that allows to completely eliminate the exposure to `Bool` in user code.

Reply:
@deepfire let's come back to the original issue I raised up - there is no need for any of the bounded instances.

The only place in current code where any of ADTs that have bounded instances are converted from or to a Bool is:

```
flag :: Flag a => a -> ArgName -> Char -> Optional HelpMessage -> Parser a
flag effect long ch help = (\case
                               True  -> effect
                               False -> opposite effect) <$> switch long ch help
```

which if you want to keep it abstract but remove the bounded:

```
flag :: Flag a => (Bool -> a) -> ArgName -> Char -> Optional HelpMessage -> Parser a
flag f long ch help = f <$> switch long ch help

toTestInstaller True = IsTestInstaller
toTestInstaller False = IsNotTestInstaller
```

and then call it as `flag toTestInstaller ...`

Reply:
Ok, let's do that.

Reply:
@domenkozar's last comment demonstrates addresses my concerns about `Bounded`, and I support moving in that direction.

Reply:
Interestingly, this is isomorphic to adding an injection `Bool -> Flag a` method to the `Flag` type class.  @domenkozar, what do you think?  It would remove the need to explicitly pass the function, and to care/remember about its name.

Reply:
cc @domenkozar

Reply:
Boolean blindness is amplified when abstractions are built around `Bool`. Since there's only one use case for the `Flag` I suggest we keep the code direct, minimizing the number of places one must look to know what happens when the flag is or is not present.

We can add more abstraction later if there's a need.

Reply:
ICYMI: https://github.com/input-output-hk/daedalus/pull/749#issuecomment-372459501

Reply:
> Boolean blindness is amplified when abstractions are built around Bool.

Which is equally true for both implementation strategies, be it:
```
instance Flag TestInstaller where
   fromBool True = IsTestInstaller
   fromBool False = IsNotTestInstaller
```
..or..
```
isTestInstallerFromBool :: Bool -> TestInstaller
isTestInstallerFromBool True = IsTestInstaller
isTestInstallerFromBool False = IsNotTestInstaller
```
right?

But I'll do as you suggest.

Reply:
Done.